cpp/include/graph.hpp

```diff
@@ -0,0 +1,217 @@
+  bool directed{false};
+  bool weighted{false};
+  bool multigraph{multigraph};
+  bool bipartite{false};
+  bool tree{false};
+  PropType has_negative_edges{PROP_UNDEF};
+  GraphProperties() = default;
+};
+
+/**
+ * @brief       Base class graphs, all but vertices and edges
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class GraphBase {
+public:
+  WT const *edge_data;     ///< edge weight
+
+  GraphProperties          prop;
+
+  VT                       number_of_vertices;
+  ET                       number_of_edges;
+
+  GraphBase(WT const *edge_data_, VT number_of_vertices_, ET number_of_edges_):
+    edge_data(edge_data_),
+    prop(),
+    number_of_vertices(number_of_vertices_),
+    number_of_edges(number_of_edges_)
+  {}
+};
+
+/**
+ * @brief       A graph stored in COO (COOrdinate) format.
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class GraphCOO: public GraphBase<VT, ET, WT> {
+public:
+  VT const *src_indices{nullptr};   ///< rowInd
+  VT const *dst_indices{nullptr};   ///< colInd
+
+  /**
+   * @brief      Default constructor
+   */
+  GraphCOO(): GraphBase<VT,ET,WT>(nullptr, 0, 0) {}
+  
+  /**
+   * @brief      Wrap existing arrays representing an edge list in a Graph.
+   *
+   *             GraphCOO does not own the memory used to represent this graph. This
+   *             function does not allocate memory.
+   *
+   * @param  source_indices        This array of size E (number of edges) contains the index of the source for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @param  destination_indices   This array of size E (number of edges) contains the index of the destination for each edge.
+   *                               Indices must be in the range [0, V-1].
+   * @param  edge_data             This array size E (number of edges) contains the weight for each edge.  This array can be null
+   *                               in which case the graph is considered unweighted.
+   * @param  number_of_vertices    The number of vertices in the graph
+   * @param  number_of_edges       The number of edges in the graph
+   */
+  GraphCOO(VT const *src_indices_, VT const *dst_indices_, WT const *edge_data_,
+           VT number_of_vertices_, ET number_of_edges_):
+    GraphBase<VT,ET,WT>(edge_data_, number_of_vertices_, number_of_edges_)
+  {}
+};
+
+/**
+ * @brief       Base class for graph stored in CSR (Compressed Sparse Row) format or CSC (Compressed Sparse Column) format
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class GraphCompressedSparseBase: public GraphBase<VT,ET,WT> {
+public:
+  VT const *offsets{nullptr};       ///< CSR offsets
```

Comment:
This should be ET, maximum offset is determined by the maximum number of edges.

Reply:
Fixed all of these.